// select-element/index.js
function selectElement(appendToID, selectID, options, selected) {

    this.appendToID = appendToID;
    this.selectID = selectID;
    this.options = options;
    this.selected = selected;

    this.selectList;

    this.create = function(cb) {
        var appendToID = document.getElementById(this.appendToID);
        this.selectList = document.createElement("select");
        this.selectList.id = this.selectID;
        appendToID.appendChild(this.selectList);
        this.update(selectID, this.options, this.selected);
    };

    this.onChange = function(cb) {
        this.selectList.addEventListener('change', function() {
            cb(this.value)
        });
    }

    this.update = function(elem, options, selected) {
        this.delete(elem);
        var selectList = document.getElementById(elem);
        for (var key in options) {
            var option = document.createElement("option");
            option.value = key;
            option.text = options[key];
            selectList.appendChild(option);

            if (key === selected) {
                option.setAttribute('selected', true);
            }
        }
    };

    this.getSelected = function(elem) {
        var selectList = document.getElementById(elem);
        var opt;
        for (var i = 0, len = selectList.options.length; i < len; i++) {
            opt = selectList.options[i];
            if (opt.selected === true) {
                return opt.value;
                break;
            }
        }
        return false;
    };

    this.delete = function(elem) {
        var selectList = document.getElementById(elem);
        for (var option in selectList) {
            selectList.remove(option);
        }
    };

    this.getAsString = function() {
        var element = document.getElementById(this.appendToID);
        var elementHtml = element.outerHTML;
        return elementHtml;
    };
}

module.exports = selectElement;
